In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split 
from sklearn.preprocessing import StandardScaler
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
import numpy as np

In [2]:
def split_scalar(X,y):
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
    sc = StandardScaler()
    X_train = sc.fit_transform(X_train)
    X_test = sc.transform(X_test)
    return X_train,X_test,y_train, y_test

def perform_lda(X,y, n):
    lda = LDA(n_components = n)
    X_train = lda.fit_transform(X, y)
    return lda, X_train
    

def logistic(X_train, y_train,X_test):
    from sklearn.linear_model import LogisticRegression
    classifier = LogisticRegression(random_state = 0)
    classifier.fit(X_train, y_train)
    cm,accuracy,classifier = cm_predition(classifier,X_test)
    return cm,accuracy,classifier

def svm_linear(X_train,y_train,X_test):
    from sklearn.svm import SVC
    classifier = SVC(kernel = 'linear', random_state = 0)
    classifier.fit(X_train, y_train)
    cm,accuracy,classifier = cm_predition(classifier,X_test)
    return  cm,accuracy,classifier
           
def svm_NL(X_train,y_train,X_test):
    from sklearn.svm import SVC
    classifier = SVC(kernel = 'rbf', random_state = 0)
    classifier.fit(X_train, y_train)
    cm,accuracy,classifier = cm_predition(classifier,X_test)
    return  cm,accuracy,classifier

def Navie(X_train,y_train,X_test):       
    # Fitting K-NN to the Training set
    from sklearn.naive_bayes import GaussianNB
    classifier = GaussianNB()
    classifier.fit(X_train, y_train)
    cm,accuracy,classifier = cm_predition(classifier,X_test)
    return  cm,accuracy,classifier         
    
    
def knn(X_train,y_train,X_test):
    # Fitting K-NN to the Training set
    from sklearn.neighbors import KNeighborsClassifier
    classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
    classifier.fit(X_train, y_train)
    cm,accuracy,classifier = cm_predition(classifier,X_test)
    return  cm,accuracy,classifier

def Decision(X_train,y_train,X_test):
    # Fitting K-NN to the Training set
    from sklearn.tree import DecisionTreeClassifier
    classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
    classifier.fit(X_train, y_train)
    cm,accuracy,classifier = cm_predition(classifier,X_test)
    return  cm,accuracy,classifier     


def random(X_train,y_train,X_test):
    # Fitting K-NN to the Training set
    from sklearn.ensemble import RandomForestClassifier
    classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
    classifier.fit(X_train, y_train)
    cm,accuracy,classifier = cm_predition(classifier,X_test)
    return  cm,accuracy,classifier

def cm_predition (classifier,X_test):
    y_pred = classifier.predict(X_test)
    from sklearn.metrics import confusion_matrix
    cm = confusion_matrix(y_test, y_pred)
    from sklearn.metrics import accuracy_score
    accuracy= accuracy_score(y_test, y_pred)
    return cm,accuracy,classifier

def lda_result(acclog,accsvmlin,accsvmnl,accnav,accknn,accdec,accrand): 
    dataframe=pd.DataFrame(index=['LDA_1'],columns=['Logistic','SVMlin','SVMNL','Navie','KNN','DT','RF'])
    for number,index in enumerate(dataframe.index):      
        dataframe['Logistic'][index]=acclog[number]
        dataframe['SVMlin'][index]=accsvmlin[number] 
        dataframe['SVMNL'][index]=accsvmnl[number] 
        dataframe['Navie'][index]=accnav[number] 
        dataframe['KNN'][index]=accknn[number] 
        dataframe['DT'][index]=accdec[number] 
        dataframe['RF'][index]=accrand[number] 
    return dataframe

In [3]:
# Importing the dataset
dataset = pd.read_csv('Wine.csv')
X = dataset.iloc[:, 0:13].values
y = dataset.iloc[:, 13].values
 
lda,X_train_lda = perform_lda(X, y, 1)
# created a empty list to append the final value
acclog=[]
accsvmlin=[]
accsvmnl=[]
accnav=[]
accknn=[]
accdec=[]
accrand=[]

X_train, X_test, y_train, y_test=split_scalar(X_train_lda,y) 

cm,accuracy,classifier=logistic(X_train,y_train,X_test)
acclog.append(accuracy)

cm,accuracy,classifier=svm_linear(X_train,y_train,X_test)
accsvmlin.append(accuracy)

cm,accuracy,classifier=svm_NL(X_train,y_train,X_test)
accsvmnl.append(accuracy)

cm,accuracy,classifier=Navie(X_train,y_train,X_test)
accnav.append(accuracy)

cm,accuracy,classifier=knn(X_train,y_train,X_test)
accknn.append(accuracy)

cm,accuracy,classifier=Decision(X_train,y_train,X_test)
accdec.append(accuracy)

cm,accuracy,classifier=random(X_train,y_train,X_test)
accrand.append(accuracy)

result = lda_result(acclog,accsvmlin,accsvmnl,accnav,accknn,accdec,accrand)
result

Logistic    SVMlin     SVMNL     Navie       KNN        DT        RF
LDA_1  0.833333  0.861111  0.833333  0.833333  0.833333  0.888889  0.833333

#1 result
         Logistic	SVMlin	      SVMNL     Navie	       KNN	        DT	       RF
LDA_1	0.833333	0.861111	0.833333	0.833333	0.833333	0.888889	0.833333

#2result 
       Logistic	 SVMlin	 SVMNL	Navie	KNN	   DT	RF
LDA_2	1.0	     1.0	 1.0	 1.0	1.0	  1.0	1.0

In [4]:
#2nd model found ok accuracy above 100%, Highest accracy in all model